### ISSUE TITLE
`describe` used by string, symbol, keyword, buffer? Docs correct?

### ISSUE THREAD

**Issue by uvtc** [janet-lang/janet#524]
The docs for `string`, `symbol`, `keyword`, and `buffer` all mention use of `describe` when necessary for converting to bytes.

But the docs for `describe` don't mention anything about bytes (but instead say it's for a human-readable description of its arg).

What `describe` previously used for something other than it's current use?

Also, what is the difference between `(string/format "%j" s)` (which I think prints out in jdn) vs `(describe s)`? JDN seems human-readable, but I think I'm missing something here.

Thanks, and thanks @sogaiu for pointing this out to me.

**Comment by sogaiu** [comment]
I had a fresh look just now and I think I may have a slightly better understanding.

The TLDR is -- perhaps the reference to `describe` is this: https://github.com/janet-lang/janet/blob/7e1d09599677c3dc5bc4ef5944e6d9e0c612e2a8/src/core/pp.c#L246-L249

The longer version is...

IIUC, `describe` is a wrapper for `janet_core_describe`:
```
static Janet janet_core_describe(int32_t argc, Janet *argv) {
    JanetBuffer *b = janet_buffer(0);
    for (int32_t i = 0; i < argc; ++i)
        janet_description_b(b, argv[i]);
    return janet_stringv(b->data, b->count);
}
```
and that calls `janet_description_b`.

`string` is a wrapper for `janet_core_string`:
```
static Janet janet_core_string(int32_t argc, Janet *argv) {
    JanetBuffer *b = janet_buffer(0);
    for (int32_t i = 0; i < argc; ++i)
        janet_to_string_b(b, argv[i]);
    return janet_stringv(b->data, b->count);
}
```

Note that it calls `janet_to_string_b` for each of the "values", where "values" is from "Values are converted to bytes if they are not byte sequences" (or the `parts` argument to `string`, etc.).

`janet_to_string_b` has a default clause: https://github.com/janet-lang/janet/blob/7e1d09599677c3dc5bc4ef5944e6d9e0c612e2a8/src/core/pp.c#L246-L249 which perhaps one might say is the "heart" of describe.  May be the defauilt clause counts as "if ... not byte sequence".

The implementations of `symbol`, `keyword`, and `buffer` are similar to that of `string`.  Thus perhaps it's a fair description of:

> Values are converted to bytes via describe if they are not byte sequences.

Possibly it might be clearer if `string`, `keyword`, `symbol`, and `buffer` are used `values` as the argument name.  At the moment, `string` uses `parts`, while the other three use `xs`.

Alternatively, perhaps the prose could be modified to match the argument names.

**Comment by uvtc** [comment]
Ok. Between your comment and @bakpakin 's, I see where my confusion was (bold at the end):

  * strings, keywords, and symbols are all just byte sequences (8-bit clean).
  * keywords have a ":" (character 58) in front of them.
  * buffers are mutable strings (again, byte sequences)
  * strings and keywords evaluate to themselves, but symbols evaluate to whatever they're bound to.
  * __when a symbol is bound to a string, both the symbol and the string it points to are byte sequences.__

The docstring for `describe` says it returns a string, which is of course (and goes without saying) a byte sequence.

The docs for `string`, `symbol`, `keyword`, and `buffer` point out they they all use `describe` to convert to bytes if necessary, and that they all concatenate.

`describe` seems to do double-duty: it not only converts to bytes but it also makes them human-readable.

Finally, the docs for `bytes?` tie everything up, saying, "Check if x is a string, symbol, keyword, or buffer."

**Comment by uvtc** [comment]
Would be nice if the docs for `describe` included what @bakpakin noted elsewhere about identity (and also about it being non-recursive):

> `describe` - Get a human readable but non-recursive description of a value as a string. This means the identity of an object should be more or less determinable from the output (i.e. if they print the same, they are probably =). For things like tables, tuples, structs and arrays, this means printing the pointer since these are recursive types.

though I'm not sure how to word it so left it out. Maybe someone else could try their hand at it?

**Comment by sogaiu** [comment]
For future reference, here is something that contains edited content from bakpakin's explanation: https://gist.github.com/sogaiu/ce9f48ccfb5ca923cc9b7b37b12cc60e

**Comment by sogaiu** [comment]
Possibly another way to keep the info about `describe` usefully "available" is to add it to janetdocs in an example or a series of examples that try to demonstrate some or all of the relevant points.
